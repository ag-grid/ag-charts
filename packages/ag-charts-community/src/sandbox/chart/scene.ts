import { createDiv, createElement, getWindow } from '../../util/dom';
import { hasConstrainedCanvasMemory } from '../../util/userAgent';
import { EventEmitter } from '../util/eventEmitter';
import type { IScene, SceneEventMap } from './types';

export class Scene implements IScene {
    static ElementClassName = 'ag-chart-wrapper';
    static ElementStyle = { position: 'relative', userSelect: 'none' };

    readonly events = new EventEmitter<SceneEventMap>();

    readonly canvas: HTMLCanvasElement;
    readonly context: CanvasRenderingContext2D;
    readonly rootElement: HTMLDivElement;
    readonly rootNode: object;

    constructor(
        public width: number,
        public height: number,
        public pixelRatio: number = getWindow('devicePixelRatio')
    ) {
        if (hasConstrainedCanvasMemory()) {
            this.pixelRatio = 1;
        }

        // Create canvas and immediately apply width + height to avoid out-of-memory errors on iOS/iPadOS Safari.
        this.canvas = createElement('canvas');
        this.setCanvasSize(width, height);

        this.context = this.canvas.getContext('2d')!;
        this.context.setTransform(this.pixelRatio, 0, 0, this.pixelRatio, 0, 0);

        this.rootElement = createDiv(Scene.ElementStyle, Scene.ElementClassName);
        this.rootElement.appendChild(this.canvas);

        this.rootNode = {};
    }

    clear() {
        this.context.save();
        this.context.setTransform(this.pixelRatio, 0, 0, this.pixelRatio, 0, 0);
        this.context.clearRect(0, 0, this.width, this.height);
        this.context.restore();
    }

    update() {}

    resize(width: number, height: number) {
        if (width === this.width && height === this.height) return;

        this.setCanvasSize(width, height);

        this.width = width;
        this.height = height;

        this.events.emit('resize', { width, height });
    }

    toDataURL(type?: string): string {
        return this.canvas.toDataURL(type);
    }

    remove() {
        this.canvas.remove();

        // Workaround memory allocation quirks in iOS Safari by resizing to 0x0 and clearing.
        // See https://bugs.webkit.org/show_bug.cgi?id=195325.
        this.canvas.width = 0;
        this.canvas.height = 0;
        this.context.clearRect(0, 0, 0, 0);

        Object.freeze(this);
    }

    private setCanvasSize(width: number, height: number) {
        const { canvas, pixelRatio } = this;

        canvas.width = Math.round(width * pixelRatio);
        canvas.height = Math.round(height * pixelRatio);

        canvas.style.width = width + 'px';
        canvas.style.height = height + 'px';
    }
}
