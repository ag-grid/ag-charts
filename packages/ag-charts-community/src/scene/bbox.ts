// For small data structs like a bounding box, objects are superior to arrays
// in terms of performance (by 3-4% in Chrome 71, Safari 12 and by 20% in Firefox 64).
// They are also self descriptive and harder to abuse.
// For example, one has to do:
// `ctx.strokeRect(bbox.x, bbox.y, bbox.width, bbox.height);`
// rather than become enticed by the much slower:
// `ctx.strokeRect(...bbox);`
// https://jsperf.com/array-vs-object-create-access

type Padding = {
    top: number;
    left: number;
    right: number;
    bottom: number;
};

export class BBox {
    x: number;
    y: number;
    width: number;
    height: number;

    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    clone() {
        const { x, y, width, height } = this;
        return new BBox(x, y, width, height);
    }

    equals(other: BBox) {
        return this.x === other.x && this.y === other.y && this.width === other.width && this.height === other.height;
    }

    containsPoint(x: number, y: number): boolean {
        return x >= this.x && x <= this.x + this.width && y >= this.y && y <= this.y + this.height;
    }

    isInfinite() {
        return (
            Math.abs(this.x) === Infinity ||
            Math.abs(this.y) === Infinity ||
            Math.abs(this.width) === Infinity ||
            Math.abs(this.height) === Infinity
        );
    }

    shrink(amounts: Partial<Padding>): this;
    shrink(amount: number, position?: 'top' | 'left' | 'bottom' | 'right' | 'vertical' | 'horizontal'): this;
    shrink(
        amount: number | Partial<Padding>,
        position?: 'top' | 'left' | 'bottom' | 'right' | 'vertical' | 'horizontal'
    ) {
        const apply = (pos: typeof position, amt: number) => {
            switch (pos) {
                case 'top':
                    this.y += amt;
                // eslint-disable-next-line no-fallthrough
                case 'bottom':
                    this.height -= amt;
                    break;
                case 'left':
                    this.x += amt;
                // eslint-disable-next-line no-fallthrough
                case 'right':
                    this.width -= amt;
                    break;
                case 'vertical':
                    this.y += amt;
                    this.height -= amt * 2;
                    break;
                case 'horizontal':
                    this.x += amt;
                    this.width -= amt * 2;
                    break;
                default:
                    this.x += amt;
                    this.width -= amt * 2;
                    this.y += amt;
                    this.height -= amt * 2;
            }
        };

        if (typeof amount === 'number') {
            apply(position, amount);
        } else {
            Object.entries(amount).forEach(([pos, amt]) => apply(pos as typeof position, amt!));
        }

        return this;
    }

    grow(amounts: Partial<Padding>): this;
    grow(amount: number, position?: 'top' | 'left' | 'bottom' | 'right' | 'vertical' | 'horizontal'): this;
    grow(
        amount: number | Partial<Padding>,
        position?: 'top' | 'left' | 'bottom' | 'right' | 'vertical' | 'horizontal'
    ) {
        if (typeof amount === 'number') {
            this.shrink(-amount, position);
        } else {
            const paddingCopy = { ...amount };

            for (const key in paddingCopy) {
                (paddingCopy as any)[key] *= -1;
            }

            this.shrink(paddingCopy);
        }

        return this;
    }

    static merge(boxes: BBox[]) {
        let left = Infinity;
        let top = Infinity;
        let right = -Infinity;
        let bottom = -Infinity;
        boxes.forEach((box) => {
            if (box.x < left) {
                left = box.x;
            }
            if (box.y < top) {
                top = box.y;
            }
            if (box.x + box.width > right) {
                right = box.x + box.width;
            }
            if (box.y + box.height > bottom) {
                bottom = box.y + box.height;
            }
        });
        return new BBox(left, top, right - left, bottom - top);
    }
}
