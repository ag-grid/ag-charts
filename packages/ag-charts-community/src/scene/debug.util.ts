import { type Node, RedrawType } from './node';

export function nodeCount(node: Node) {
    let count = 1;
    let visibleCount = node.visible ? 1 : 0;
    let dirtyCount = node.dirty > RedrawType.NONE ? 1 : 0;

    for (const child of node.children(false)) {
        const c = nodeCount(child);
        count += c.count;
        dirtyCount += c.dirtyCount;
        visibleCount += c.visibleCount;
    }

    return { count, visibleCount, dirtyCount };
}
