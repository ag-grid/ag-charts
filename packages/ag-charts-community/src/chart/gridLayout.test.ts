import { describe, expect, test } from '@jest/globals';

import { BBox } from '../scene/bbox';
import { gridLayout } from './gridLayout';

describe('horizontal layout', () => {
    test('one row, multiple columns, one page', () => {
        const MAX_HEIGHT = 23;
        const MAX_WIDTH = 510;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'horizontal',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two rows, multiple columns, one page', () => {
        const MAX_HEIGHT = 47;
        const MAX_WIDTH = 510;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'horizontal',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two rows, multiple columns, one page, mixed item widths in single column ', () => {
        const MAX_HEIGHT = 47;
        const MAX_WIDTH = 510;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 100, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'horizontal',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two rows, multiple columns, two pages', () => {
        const MAX_HEIGHT = 47;
        const MAX_WIDTH = 510;
        const BBOXES: BBox[] = [
            // page 1
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),

            // page 2
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'horizontal',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two rows, multiple columns, two page, mixed item widths in single column ', () => {
        const MAX_HEIGHT = 47;
        const MAX_WIDTH = 510;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 100, 23),
        ];

        // 100, 120, 140, 140
        // 140, 140, 120, 100

        // 100, 120, 140,
        // 140, 140, 120,

        // 120, 100

        const { pages } = gridLayout({
            orientation: 'horizontal',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });
});

describe('vertical layout', () => {
    test('one column, multiple rows, one page', () => {
        const MAX_HEIGHT = 100;
        const MAX_WIDTH = 150;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'vertical',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two columns, multiple rows, one page', () => {
        const MAX_HEIGHT = 100;
        const MAX_WIDTH = 290;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'vertical',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two columns, multiple rows, one page, mixed item widths in single column ', () => {
        const MAX_HEIGHT = 100;
        const MAX_WIDTH = 290;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 100, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'vertical',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two columns, multiple rows, two pages', () => {
        const MAX_HEIGHT = 100;
        const MAX_WIDTH = 290;
        const BBOXES: BBox[] = [
            // page 1
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),

            // page 2
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 140, 23),
            new BBox(0, 0, 140, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'vertical',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });

    test('two columns, multiple rows, two page, mixed item widths in single column ', () => {
        const MAX_HEIGHT = 100;
        const MAX_WIDTH = 290;
        const BBOXES: BBox[] = [
            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 100, 23),

            new BBox(0, 0, 100, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 130, 23),
            new BBox(0, 0, 125, 23),
            new BBox(0, 0, 120, 23),
            new BBox(0, 0, 100, 23),
        ];

        const { pages } = gridLayout({
            orientation: 'vertical',
            bboxes: BBOXES,
            maxHeight: MAX_HEIGHT,
            maxWidth: MAX_WIDTH,
        });

        expect(pages).toMatchSnapshot();
    });
});
