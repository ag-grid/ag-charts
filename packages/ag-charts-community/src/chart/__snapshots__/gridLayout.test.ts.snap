// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`horizontal layout one row, multiple columns, one page 1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 23,
        "columnWidth": 100,
        "indices": [
          0,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 23,
        "columnWidth": 120,
        "indices": [
          1,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 23,
        "columnWidth": 140,
        "indices": [
          2,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 23,
        "columnWidth": 140,
        "indices": [
          3,
        ],
      },
    ],
    "endIndex": 3,
    "pageHeight": 23,
    "pageWidth": 500,
    "startIndex": 0,
  },
]
`;

exports[`horizontal layout two rows, multiple columns, one page 1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 100,
        "indices": [
          0,
          4,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 120,
        "indices": [
          1,
          5,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          2,
          6,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          3,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 46,
    "pageWidth": 500,
    "startIndex": 0,
  },
]
`;

exports[`horizontal layout two rows, multiple columns, one page, mixed item widths in single column  1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 130,
        "indices": [
          0,
          4,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 125,
        "indices": [
          1,
          5,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 125,
        "indices": [
          2,
          6,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 130,
        "indices": [
          3,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 46,
    "pageWidth": 510,
    "startIndex": 0,
  },
]
`;

exports[`horizontal layout two rows, multiple columns, two page, mixed item widths in single column  1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          0,
          3,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          1,
          4,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          2,
          5,
        ],
      },
    ],
    "endIndex": 5,
    "pageHeight": 46,
    "pageWidth": 420,
    "startIndex": 0,
  },
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 23,
        "columnWidth": 120,
        "indices": [
          6,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 23,
        "columnWidth": 100,
        "indices": [
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 23,
    "pageWidth": 220,
    "startIndex": 6,
  },
]
`;

exports[`horizontal layout two rows, multiple columns, two pages 1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 100,
        "indices": [
          0,
          4,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 120,
        "indices": [
          1,
          5,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          2,
          6,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          3,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 46,
    "pageWidth": 500,
    "startIndex": 0,
  },
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 100,
        "indices": [
          8,
          12,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 120,
        "indices": [
          9,
          13,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          10,
          14,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 46,
        "columnWidth": 140,
        "indices": [
          11,
          15,
        ],
      },
    ],
    "endIndex": 15,
    "pageHeight": 46,
    "pageWidth": 500,
    "startIndex": 8,
  },
]
`;

exports[`vertical layout one column, multiple rows, one page 1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          0,
          1,
          2,
          3,
        ],
      },
    ],
    "endIndex": 3,
    "pageHeight": 92,
    "pageWidth": 140,
    "startIndex": 0,
  },
]
`;

exports[`vertical layout two columns, multiple rows, one page 1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          0,
          1,
          2,
          3,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          4,
          5,
          6,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 92,
    "pageWidth": 280,
    "startIndex": 0,
  },
]
`;

exports[`vertical layout two columns, multiple rows, one page, mixed item widths in single column  1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 130,
        "indices": [
          0,
          1,
          2,
          3,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 130,
        "indices": [
          4,
          5,
          6,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 92,
    "pageWidth": 260,
    "startIndex": 0,
  },
]
`;

exports[`vertical layout two columns, multiple rows, two page, mixed item widths in single column  1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 130,
        "indices": [
          0,
          1,
          2,
          3,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 130,
        "indices": [
          4,
          5,
          6,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 92,
    "pageWidth": 260,
    "startIndex": 0,
  },
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 130,
        "indices": [
          8,
          9,
          10,
          11,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 130,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 125,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 130,
        "indices": [
          12,
          13,
          14,
          15,
        ],
      },
    ],
    "endIndex": 15,
    "pageHeight": 92,
    "pageWidth": 260,
    "startIndex": 8,
  },
]
`;

exports[`vertical layout two columns, multiple rows, two pages 1`] = `
[
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          0,
          1,
          2,
          3,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          4,
          5,
          6,
          7,
        ],
      },
    ],
    "endIndex": 7,
    "pageHeight": 92,
    "pageWidth": 280,
    "startIndex": 0,
  },
  {
    "columns": [
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          8,
          9,
          10,
          11,
        ],
      },
      {
        "bboxes": [
          BBox {
            "height": 23,
            "width": 100,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 120,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
          BBox {
            "height": 23,
            "width": 140,
            "x": 0,
            "y": 0,
          },
        ],
        "columnHeight": 92,
        "columnWidth": 140,
        "indices": [
          12,
          13,
          14,
          15,
        ],
      },
    ],
    "endIndex": 15,
    "pageHeight": 92,
    "pageWidth": 280,
    "startIndex": 8,
  },
]
`;
